pkg/proc/threads.go

```diff
@@ -209,6 +209,24 @@ func next(dbp Process, stepInto, inlinedStepOut bool) error {
 				continue
 			}
 
+			// On 386 linux when buildmode is pie, compiler insert more instructions (ex: call __x86.get_pc_thunk.).
```

Comment:
I think this is too hacky, it would be preferrable to figure out the address of x86.get_pc_thunk by reading the symbol table and add an entry for that in bi.Functions.

Reply:
En, I have try to get symbol from symbol table by the addr calleld, it is ok. 

----

But  the comments of `Functions` is
`Functions is a list of all DW_TAG_subprogram entries in debug_info, sorted by entry point`. 

I am not sure additional funcs from symbol table, whether destroy the original logic of `bi.Functions`(addtional funcs has no fn.entry, fn.end, dwarfoffset and compileUnit, just has name and addr if my understanding is right).  

I think other way is  add one filed named `SymNames`( type is `map[addr]elf.Symbol`). This map is just used by  `bi.symLookup`

Reply:
True, it would probably be a problem to have Function objects without a compile unit. 

Instead of keeping around the whole gosym.Table map I'd just use `gosym.(*Table).PCToFunc("__x86.get_pc_thunk")` to get its entry point and then the check is `instr.IsCall() && instr.DestLoc.PC == entryPointOfx86getPCThunk`.

Reply:
1.
__x86.get_pc_thunk is a array, even same name but in diffrent place.

```
addr       name
127520 __x86.get_pc_thunk.ax
127530 __x86.get_pc_thunk.cx
127540 __x86.get_pc_thunk.dx
127550 __x86.get_pc_thunk.bx
127560 __x86.get_pc_thunk.bp
127570 __x86.get_pc_thunk.si
127580 __x86.get_pc_thunk.di
12750c __x86.get_pc_thunk.dx
127510 __x86.get_pc_thunk.bx
``` 
Can't not use entryPointOfx86getPCThunk.   

-----

2. 
And one benefit for addtional field SymNames in `bi` to help search name.

diassemble
```
(dlv) disassemble
TEXT main.Foo(SB) /root/workspace/go-delve/delve/_fixtures/teststepconcurrent.go
        teststepconcurrent.go:13        0x1c0eb0        e81b6ef7ff              call $C._IO_2_1_stdout_+163024
        teststepconcurrent.go:13        0x1c0eb5        c7c1fcffffff            mov ecx, 0xfffffffc
        teststepconcurrent.go:13        0x1c0ebb        658b09                  mov ecx, dword ptr gs:[ecx]
        teststepconcurrent.go:13        0x1c0ebe        3b6108                  cmp esp, dword ptr 
```

will become

```
(dlv) disassemble
TEXT main.Foo(SB) /root/workspace/go-delve/delve/_fixtures/teststepconcurrent.go
        teststepconcurrent.go:13        0x1c0eb0        e81b6ef7ff              call $__x86.get_pc_thunk.cx
        teststepconcurrent.go:13        0x1c0eb5        c7c1fcffffff            mov ecx, 0xfffffffc
        teststepconcurrent.go:13        0x1c0ebb        658b09                  mov ecx, dword ptr gs:[ecx]
        teststepconcurrent.go:13        0x1c0ebe        3b6108                  cmp esp, dword ptr [ecx+0x8]

```  
---- 

3. 
I find that here should be clean for `theard.go`. 
Place function `bi.IsX86GetPCThunk(pc uint64)`  into bi. 
And just keep clean here.
```
if instr.IsCall() && instr.DestLoc != nil && dbp.BinInfo().IsX86GetPCThunk(instr.DestLoc.PC) {
     continue
}
```

Reply:
I don't want to depend on the assembly formatter behaving in a certain way to implement an unrelated feature. And the `IsX86GetPCThunk` should be called `InhibitStepInto` (or something like that) and be part of the Arch object, probably.

Reply:
I agree with you and get it.

Reply:
As I was saying before the test calling `bi.Arch.InhibitStepInto(bi, instr.DestLoc.PC)` should go inside setStepIntoBreakpoint, remove the check for `instr.DestLoc.Fn != nil` in the if at line 216 and inside setStepIntoBreakpoint do:

```
if dbp.BinInfo().Arch.InhibitStepInto(dbp.BinInfo(), instr.DestLoc.PC) {
    return nil
}
```

around the same place where we skip private runtime functions.

Reply:
Ok, done and rebase.